onyo/argparse_helpers.py

```diff
@@ -18,26 +32,39 @@ def __call__(self,
                  namespace: argparse.Namespace,
                  key_values: list[str],
                  option_string: str | None = None) -> None:
-        """Turn a list of 'key=value' pairs into a list of dictionaries
+        r"""Turn a list of 'KEY=VALUE' pairs into a list of dictionaries.
+
+        Each KEY can be defined either 1 or N times (where N is the number of
+        dictionaries to be created).
 
-        Every key appearing multiple times in `key=value` is applied to a new dictionary every time.
-        All keys appearing multiple times, must appear the same number of times (and thereby define the number of dicts
-        to be created). In case of different counts: raise.
-        Every key appearing once in `key_values` will be applied to all dictionaries.
+        A KEY that is declared once will apply to all dictionaries.
+
+        All KEYs appearing N times must appear the same number of times. If not,
+        a message will print to standard error and the program will exit with
+        status code 2.
+
+        Parameters
+        ----------
+        parser
+        namespace
+        key_values
+            List of strings containing key-value pairs.
+        option_string
         """
 
         for kv in key_values:
             if "=" not in kv:
                 parser.error(f"Invalid argument '{kv}'. Expected key-value pairs '<key>=<value>'.")
+
         pairs = [p.split('=', maxsplit=1) for p in key_values]
-        register_dict = {k: [] for k, v in pairs}
-        [register_dict[k].append(v) for k, v in pairs]
-        number_of_dicts = max(len(v) for v in register_dict.values())
-        invalid_keys = [(k, len(v)) for k, v in register_dict.items() if 1 < len(v) < number_of_dicts]
-        if invalid_keys:
-            parser.error(f"All keys given multiple times must be provided the same number of times."
-                         f"Max. times a key was given: {number_of_dicts}.{os.linesep}"
-                         f"But also got: {', '.join(['{} {} times'.format(k, c) for k, c in invalid_keys])}")
+        key_lists = {k: [] for k, v in pairs}
+        [key_lists[k].append(v) for k, v in pairs]
+
+        key_counts = {k: len(v) for k, v in key_lists.items()}
+        key_max_count = max(key_counts.values())
+        if any([True for k, c in key_counts.items() if 1 < c < key_max_count]):
+            parser.error(f"All keys given multiple times must be given the same number of times:{os.linesep}"
```

Comment:
Seeing this, I feel like `parser.error()` should possibly replaced with `ui.error()` -- except obviously if it does something special and there is a reason for the usage of this function.

At least, when I began to apply `ui.error()` and `ui.print()` throughout the code, I just looked for normal prints, so it might have slipped through earlier.

Reply:
Nope. This is telling argparse to error out before anything onyo is called. It would report what's wrong with the call itself.